extension/goplugin/schemas.go

```diff
@@ -0,0 +1,597 @@
+
+	if err != nil {
+		return err
+	}
+	resourceType, ok := GlobRawTypes[thisSchema.rawSchema.ID]
+	if !ok {
+		return fmt.Errorf("No type registered for schema title: %s", thisSchema.rawSchema.ID)
+	}
+	resource := reflect.ValueOf(res)
+
+	for i := 0; i < resourceType.NumField(); i++ {
+		field := resource.Elem().Field(i)
+
+		fieldType := resourceType.Field(i)
+		propertyName := fieldType.Tag.Get("db")
+		property, err := thisSchema.rawSchema.GetPropertyByID(propertyName)
+		if err != nil {
+			return err
+		}
+		value := data[property.ID]
+		thisSchema.assignField(propertyName, field, value)
+	}
+
+	return nil
+}
+
+func setValue(field, value reflect.Value) {
+	if value.IsValid() {
+		if value.Type() != field.Type() && field.Kind() == reflect.Slice { // empty slice has type []interface{}
+			field.Set(reflect.MakeSlice(field.Type(), 0, 0))
+		} else {
+			field.Set(value)
+		}
+	}
+}
+
+// LockFetch locks and fetches resource by ID
+func (thisSchema *Schema) LockFetch(id string, res interface{}, context goext.Context, policy goext.LockPolicy) error {
+	return thisSchema.Fetch(id, res, context)
+}
+
+type TransactionData struct {
+	Tx    transaction.Transaction
+	Depth int
+}
+
+// Create creates a resource
+func (thisSchema *Schema) Create(resource interface{}, context goext.Context) error {
+	if !isPointer(resource) {
+		return NotPointerErr
+	}
+
+	if context == nil {
+		context = goext.MakeContext()
+	}
+	tx, hasOpenTransaction := contextGetTransaction(context)
+	if hasOpenTransaction {
+		contextCopy := goext.MakeContext().
+			WithSchemaID(thisSchema.ID()).
+			WithResource(thisSchema.structToMap(resource))
+		contextSetTransaction(contextCopy, tx)
+		return thisSchema.createInTransaction(resource, contextCopy, tx)
+	}
+
+	context.WithSchemaID(thisSchema.ID()).
+		WithResource(thisSchema.structToMap(resource))
+
+	if err := thisSchema.environment.HandleEvent(goext.PreCreate, context); err != nil {
+		return err
+	}
+
+	tx, err := thisSchema.environment.Database().Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Close()
+	contextSetTransaction(context, tx)
+
+	if err = thisSchema.environment.HandleEvent(goext.PreCreateTx, context); err != nil {
+		return err
+	}
+
+	if err = tx.Create(thisSchema, context["resource"].(map[string]interface{})); err != nil {
+		return err
+	}
+
+	if err = thisSchema.environment.updateResourceFromContext(resource, context); err != nil {
+		return err
+	}
+
+	if err = thisSchema.environment.HandleEvent(goext.PostCreateTx, context); err != nil {
+		return err
+	}
+
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+
+	if err = tx.Close(); err != nil {
+		return err
+	}
+
+	return thisSchema.environment.HandleEvent(goext.PostCreate, context)
+}
+
+func (thisSchema *Schema) createInTransaction(resource interface{}, context goext.Context, tx goext.ITransaction) error {
+	var err error
+
+	if err := thisSchema.environment.HandleEvent(goext.PreCreate, context); err != nil {
```

Comment:
this function is (by definition) executed in transaction, so why are the Pre/Post handlers invoked?